src/Common/src/Microsoft/Win32/SafeHandles/NtlmSecuritySafeHandles.cs

```diff
@@ -0,0 +1,271 @@
+            const int HMacDigestLength = 8;
+
+            byte[] output = new byte[Interop.NetSecurity.MD5DigestLength];
+            Array.Clear(output, 0, output.Length);
+            MarshalUint(output, 0, Version); // version
+            MarshalUint(output, SequenceNumberOffset, _sequenceNumber);
+            byte[] hash;
+
+            using (var incremental = IncrementalHash.CreateHMAC(HashAlgorithmName.MD5, _digest))
+            {
+                incremental.AppendData(output, SequenceNumberOffset, ChecksumOffset);
+                incremental.AppendData(buffer, offset, count);
+                hash = incremental.GetHashAndReset();
+                _sequenceNumber++;
+            }
+
+            if (sealingKey == null)
+            {
+                Array.Copy(hash, 0, output, ChecksumOffset, HMacDigestLength);
+            }
+            else
+            {
+                byte[] cipher = sealingKey.SealOrUnseal(true, hash, 0, HMacDigestLength);
+                Array.Copy(cipher, 0, output, ChecksumOffset, cipher.Length);
+            }
+
+            return output;
+        }
+
+        public byte[] SealOrUnseal(bool seal, byte[] buffer, int offset, int count)
+        {
+            //Message Confidentiality. Reference: https://msdn.microsoft.com/en-us/library/cc236707.aspx
+            Debug.Assert(_isSealingKey, "Cannot seal or unseal with signing key");
+            Debug.Assert(offset >= 0 && offset < buffer.Length, "Cannot sign with invalid offset " + offset);
+            Debug.Assert((count + offset) <= buffer.Length, "Cannot sign with invalid count ");
+
+            unsafe
+            {
+                fixed (byte* bytePtr = buffer)
+                {
+                    // Since RC4 is XOR-based, encrypt or decrypt is relative to input data
+                    // reference: https://msdn.microsoft.com/en-us/library/cc236707.aspx
+                    byte[] output = new byte[count];
+
+                    Interop.Crypto.EvpCipher(_cipherContext, output, (bytePtr + offset), count);
+                    return  output;
+                }
+            }
+        }
+
+        private static void MarshalUint(byte[] ptr, int offset, uint num)
+        {
+            for (int i = 0; i < 4; i++)
+            {
+                ptr[offset + i] = (byte) (num & 0xff);
+                num >>= 8;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a ntlm_type2*
+    /// </summary>
+    internal sealed class SafeNtlmType2Handle : SafeHandle
+    {
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurity.HeimNtlmFreeType2(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+
+        private SafeNtlmType2Handle() : base(IntPtr.Zero, true)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a ntlm_type3*
```

Comment:
The comment and name say Type 3 (whatever that means).  The code appears to be entirely about Type 2.